src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs

```diff
@@ -1251,4 +1251,1192 @@ static class C
+            //         Out(in p);
+            Diagnostic(ErrorCode.ERR_BadArgRef, "p").WithArguments("1", "out").WithLocation(12, 16));
+    }
+
+    [Fact(Skip = "https://github.com/dotnet/roslyn/issues/68714")]
+    public void RefReadonlyParameter_Arglist()
+    {
+        var source = """
+            class C
+            {
+                static void M(ref readonly int p, __arglist) => System.Console.WriteLine(p);
+                static void Main()
+                {
+                    int x = 111;
+                    M(x, __arglist(x));
+                    M(ref x, __arglist(x));
+                    M(in x, __arglist(x));
+                }
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "111");
+        verifier.VerifyDiagnostics(
+            // (7,11): warning CS9503: Argument 1 should be passed with 'ref' or 'in' keyword
+            //         M(x, __arglist(x));
+            Diagnostic(ErrorCode.WRN_ArgExpectedRefOrIn, "x").WithArguments("1").WithLocation(7, 11));
+    }
+
+    [Fact]
+    public void RefReadonlyParameter_Arglist_OutArgument()
+    {
+        var source = """
+            class C
+            {
+                static void M(ref readonly int p, __arglist) => System.Console.WriteLine(p);
+                static void Main()
+                {
+                    int x = 111;
+                    M(out x, __arglist(x));
+                }
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (7,15): error CS1615: Argument 1 may not be passed with the 'out' keyword
+            //         M(out x, __arglist(x));
+            Diagnostic(ErrorCode.ERR_BadArgExtraRef, "x").WithArguments("1", "out").WithLocation(7, 15));
+    }
+
+    [Fact]
+    public void RefReadonlyParameter_CrossAssembly()
+    {
+        var source1 = """
+            public class C
+            {
+                public void M(ref readonly int p) => System.Console.Write(p);
+                void M2()
+                {
+                    int x = 5;
+                    M(x);
+                    M(ref x);
+                    M(in x);
+                }
+            }
+            """;
+        var comp1 = CreateCompilation(source1).VerifyDiagnostics(
+            // (7,11): warning CS9503: Argument 1 should be passed with 'ref' or 'in' keyword
+            //         M(x);
+            Diagnostic(ErrorCode.WRN_ArgExpectedRefOrIn, "x").WithArguments("1").WithLocation(7, 11));
+
+        var source2 = """
+            class D
+            {
+                void M(C c)
+                {
+                    int x = 6;
+                    c.M(x);
+                    c.M(ref x);
+                    c.M(in x);
+                }
+            }
+            """;
+        CreateCompilation(source2, new[] { comp1.ToMetadataReference() }, parseOptions: TestOptions.Regular11).VerifyDiagnostics();
```

Comment:
I think no error is better, although perhaps we should add some warnings.

The design of this feature is based on backward compatibility. And .NET 8 wants to introduce `ref readonly` parameters. But reporting errors here would mean users would need to change their LangVersion to 12 which I think isn't otherwise required, so migrating to `ref readonly` in .NET 8 would be a breaking change.

I think this is not a problem for `C#11` compiler (which will report errors here), because that cannot be used with .NET 8. But I might be wrong (about both cases).

Reply:
Okay, after offline discussion, I will implement your suggestion. It's probably best that LangVersion=11 and old compilers are consistent here.